lib/SIL/SILFunctionType.cpp

```diff
@@ -2143,10 +2143,17 @@ namespace {
 
     // SIL type lowering only does special things to tuples and functions.
 
-    /// Functions need to preserve their abstraction structure.
-    CanSILFunctionType visitSILFunctionType(CanSILFunctionType origType,
-                                            bool dropGenerics = false)
-    {
+    // When a function appears inside of another type, we only perform
+    // substitutions if it does not have a generic signature.
+    CanSILFunctionType visitSILFunctionType(CanSILFunctionType origType) {
+      if (origType->getGenericSignature())
+        return origType;
```

Comment:
No. Think of it like a lambda expression. If I have `x + 2` and a I substitute `x := 1`, the result is `1 + 2`. But if I have `(lambda (x) (x + 2))`, and I substitute `x := 1`, the result does not change because the `x` inside the lambda is a "new" term.